### PR TITLE
Output matrix and util/matrix update

### PR DESCRIPTION
**Pull request by MartinKarp**
Added output matrix to evd_tol cyclic and classic aswell as updated the functions in the util/matrix file.

### CODE REVIEW

File: src/evd/classic/evd.cpp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #include "evd.hpp"
 #include <math.h>
+#include <stdio.h>
```
Comment (RajaSimha) [inline src/evd/classic/evd.cpp]: Remove this. I guess `stdio.h` is not necessary.

Comment (MartinKarp) [reply src/evd/classic/evd.cpp]: Done!

**Comment by eozd** [comment]
Thank you for the interface update. I just checked the EVD algorithm a little in-depth. It looks like the updates you perform for data matrix and the eigenvector matrix are nearly identical. To me, it seems like there is some extra overhead here. Can't you just compute the eigenvector matrix without explicitly diagonalizing the input matrix?

Let us call the input matrix `X`. I write `X = ULU^T`. Then, `L = U^TXU`. Since `L` is diagonal, computing `i`th eigenvalue which is equal to `L_ii` can be done as `x_ii * dot(u_i, u_i)` where `u_i` is the `i`th column of `U`.

**Comment by MartinKarp** [comment]
> Thank you for the interface update. I just checked the EVD algorithm a little in-depth. It looks like the updates you perform for data matrix and the eigenvector matrix are nearly identical. To me, it seems like there is some extra overhead here. Can't you just compute the eigenvector matrix without explicitly diagonalizing the input matrix?
> 
> Let us call the input matrix `X`. I write `X = ULU^T`. Then, `L = U^TXU`. Since `L` is diagonal, computing `i`th eigenvalue which is equal to `L_ii` can be done as `x_ii * dot(u_i, u_i)` where `u_i` is the `i`th column of `U`. Or is this analysis flawed?
> 
> Edit: The above formulas are incorrect. `L_ii = dot(u_i, dot(X, u_i))` is the correct one I believe.

Yes, I investigated this earlier this week, however my conclusion is that the update `L_ii = dot(u_i, dot(X, u_i))` requires `O(n^2)` ops while the update we on matrix X is  `O(n)` per iteration. This is why we compute the update explicitly each iteration and do not just get the values through a product of X and the eigenvector matrix. Does this answer your concerns or did I miss the point?

**Comment by RajaSimha** [comment]
> Thank you for the interface update. I just checked the EVD algorithm a little in-depth. It looks like the updates you perform for data matrix and the eigenvector matrix are nearly identical. To me, it seems like there is some extra overhead here. Can't you just compute the eigenvector matrix without explicitly diagonalizing the input matrix?
> 
> Let us call the input matrix `X`. I write `X = ULU^T`. Then, `L = U^TXU`. Since `L` is diagonal, computing `i`th eigenvalue which is equal to `L_ii` can be done as `x_ii * dot(u_i, u_i)` where `u_i` is the `i`th column of `U`. Or is this analysis flawed?
> 
> Edit: The above formulas are incorrect. `L_ii = dot(u_i, dot(X, u_i))` is the correct one I believe.

I guess it is not possible because of the following reason (As per my understanding of your question):

Updating data matrix is essential because that determines the new values of rotation parameters - `sin(x)` and `cos(x)` required to update the eigenvector matrix in the next iteration since there exists a dependency on the (updated) data matrix.

**Comment by eozd** [comment]
> > Thank you for the interface update. I just checked the EVD algorithm a little in-depth. It looks like the updates you perform for data matrix and the eigenvector matrix are nearly identical. To me, it seems like there is some extra overhead here. Can't you just compute the eigenvector matrix without explicitly diagonalizing the input matrix?
> > Let us call the input matrix `X`. I write `X = ULU^T`. Then, `L = U^TXU`. Since `L` is diagonal, computing `i`th eigenvalue which is equal to `L_ii` can be done as `x_ii * dot(u_i, u_i)` where `u_i` is the `i`th column of `U`. Or is this analysis flawed?
> > Edit: The above formulas are incorrect. `L_ii = dot(u_i, dot(X, u_i))` is the correct one I believe.
> 
> Yes, I investigated this earlier this week, however my conclusion is that the update `L_ii = dot(u_i, dot(X, u_i))` requires `O(n^2)` ops while the update we on matrix X is `O(n)` per iteration. This is why we compute the update explicitly each iteration and do not just get the values through a product of X and the eigenvector matrix. Does this answer your concerns or did I miss the point?

You don't need to compute eigenvalues at each iteration. They are only needed as a return value.

**Comment by eozd** [comment]
> > Thank you for the interface update. I just checked the EVD algorithm a little in-depth. It looks like the updates you perform for data matrix and the eigenvector matrix are nearly identical. To me, it seems like there is some extra overhead here. Can't you just compute the eigenvector matrix without explicitly diagonalizing the input matrix?
> > Let us call the input matrix `X`. I write `X = ULU^T`. Then, `L = U^TXU`. Since `L` is diagonal, computing `i`th eigenvalue which is equal to `L_ii` can be done as `x_ii * dot(u_i, u_i)` where `u_i` is the `i`th column of `U`. Or is this analysis flawed?
> > Edit: The above formulas are incorrect. `L_ii = dot(u_i, dot(X, u_i))` is the correct one I believe.
> 
> I guess it is not possible because of the following reason (As per my understanding of your question):
> 
> Updating data matrix is essential because that determines the new values of rotation parameters - `sin(x)` and `cos(x)` required to update the eigenvector matrix in the next iteration since there exists a dependency on the (updated) data matrix.

Yeah, I think this is the real reason why we need `A`. Maybe `sin` and `cos` values can be computed by `Q` as well, but I couldn't find a way to do it.

**Pull request merged**